### PR TITLE
add hmc6300 and hmc6301 driver

### PR DESCRIPTION
**Pull request by buha**
HMC6300 is a millimeterwave transmitter within 57 and 64 GHz.
HMC6301 is a millimeterwave receiver within 57 and 64 GHz.

This initial driver version supports reading/writing registers,
reading/writing the whole register map, setting a few specific
parameters and setting the VCO frequency and making sure the
PLL is locked.

### CODE REVIEW

File: drivers/rf-transceiver/hmc6300/hmc6300.c
Comment (buha) [inline drivers/rf-transceiver/hmc6300/hmc6300.c:407]: gotta remove these comments

File: drivers/rf-transceiver/hmc6300/hmc6300.c
Comment (amiclaus) [inline drivers/rf-transceiver/hmc6300/hmc6300.c:358]: same here: we can use the write function inline instead for adding extra wrapper.

**Review by amiclaus — COMMENTED** [review]
some comments on my side.

File: drivers/rf-transceiver/hmc6300/hmc6300.h
Comment (buha) [reply drivers/rf-transceiver/hmc6300/hmc6300.h:56]: So I've intentionally not used alignment because alignment is easily screwed by editor settings. 1 whitespace will never be displayed incorrectly.

File: drivers/rf-transceiver/hmc6300/hmc6300.h
Comment (buha) [reply drivers/rf-transceiver/hmc6300/hmc6300.h:41]: I've skipped these, they are not really helpful, I've checked linux drivers, they don't have them either.

File: drivers/rf-transceiver/hmc6300/hmc6300.c
Comment (buha) [reply drivers/rf-transceiver/hmc6300/hmc6300.c:353]: They are not really needed but they are helpful I guess, IF/RF attenuation is what users want to change and this naming avoids some looking up through register map... some registers have some weird names on this chip.

File: drivers/rf-transceiver/hmc6300/hmc6300.h
Comment (buha) [inline drivers/rf-transceiver/hmc6300/hmc6300.h:48]: setting->param

File: drivers/rf-transceiver/hmc6300/hmc6300.c
Comment (CiprianRegus) [inline drivers/rf-transceiver/hmc6300/hmc6300.c:202]: The readability may be improved by using `no_os_field_prep()` and macros for the magic values. Also, the casts are not needed.

File: drivers/rf-transceiver/hmc6300/hmc6300.c
Comment (CiprianRegus) [inline drivers/rf-transceiver/hmc6300/hmc6300.c:205]: `18` might be replaced with `HMC6300_FRAME_SIZE` or a similar macro.

**Review by CiprianRegus — COMMENTED** [review]
Also, the commit message seems to have some squash left overs.

File: drivers/rf-transceiver/hmc6300/hmc6300.c
Comment (buha) [reply drivers/rf-transceiver/hmc6300/hmc6300.c:233]: I think here it's acceptable to not check the return because it can only be 0 or -EINVAL, and we know through code review that we're passing valid parameters to set_value function. I will add an in-code comment about this.

File: drivers/rf-transceiver/hmc6300/hmc6300.c
Comment (buha) [reply drivers/rf-transceiver/hmc6300/hmc6300.c:182]: Hmm, I was also wondering whether I should do it or not. To me, if a remove() fails, so be it, but we need to remove the others too, otherwise we end up with memory leaks.

File: drivers/rf-transceiver/hmc6300/hmc6300.c
Comment (buha) [reply drivers/rf-transceiver/hmc6300/hmc6300.c:229]: good point, will do !

File: drivers/rf-transceiver/hmc6300/hmc6300.c
Comment (buha) [reply drivers/rf-transceiver/hmc6300/hmc6300.c:202]: will do

**Comment by buha** [comment]
V3:
- addressed the remainder of the feedback

**Comment by buha** [comment]
Ok so I'm going to add hmc6301 support in this same PR, let's not merge until I have that ready.

**Review by amiclaus — COMMENTED** [review]
> Ok so I'm going to add hmc6301 support in this same PR, let's not merge until I have that ready.

I'd suggest changing it to draft.

**Comment by buha** [comment]
V4:
- added hmc6301 support

**Comment by buha** [comment]
V5: 
- fixed commit message

File: drivers/rf-transceiver/hmc630x/hmc630x.c
Comment (CiprianRegus) [inline drivers/rf-transceiver/hmc630x/hmc630x.c:657]: Wouldn't `-ENOSYS` fit here?

File: drivers/rf-transceiver/hmc630x/hmc630x.h
Comment (CiprianRegus) [inline drivers/rf-transceiver/hmc630x/hmc630x.h:270]: Why are these defined in the source file instead of here?

Comment (buha) [reply drivers/rf-transceiver/hmc630x/hmc630x.h:270]: I guess it makes more sense to hide these from the application so that apps perform any read/write through API. This avoids situations where API users mess with device.parameter directly.

**Comment by buha** [comment]
V6:
- ENOSYS where appropriate instead of EINVAL

**Review by CiprianRegus — APPROVED** [review]

**Comment by buha** [comment]
Got feedback about some bit inversions in the datasheet which are not handled properly. Will implement that and reask for review.

**Comment by buha** [comment]
V7:
- some bitfields are inverted in memory so I've inverted the values in the enums, should be good to merge now if no more comments.

File: drivers/rf-transceiver/hmc630x/hmc630x.c
Comment (amiclaus) [inline drivers/rf-transceiver/hmc630x/hmc630x.c:100]: nit: `for (` 😃

File: drivers/rf-transceiver/hmc630x/hmc630x.c
Comment (amiclaus) [inline drivers/rf-transceiver/hmc630x/hmc630x.c:143]: nit: `switch (`

File: drivers/rf-transceiver/hmc630x/hmc630x.c
Comment (amiclaus) [inline drivers/rf-transceiver/hmc630x/hmc630x.c:530]: nit: i'd separate variable declaration from the rest of the code with a new line

File: drivers/rf-transceiver/hmc630x/hmc630x.c
Comment (amiclaus) [inline drivers/rf-transceiver/hmc630x/hmc630x.c:531]: are we 100% sure that `dev->type` is set correctly (either HMC6301 or HMC6300)?

File: drivers/rf-transceiver/hmc630x/hmc630x.c
Comment (buha) [reply drivers/rf-transceiver/hmc630x/hmc630x.c]: fixed

File: drivers/rf-transceiver/hmc630x/hmc630x.c
Comment (buha) [reply drivers/rf-transceiver/hmc630x/hmc630x.c]: fixed

File: drivers/rf-transceiver/hmc630x/hmc630x.c
Comment (buha) [reply drivers/rf-transceiver/hmc630x/hmc630x.c:530]: fixed

File: drivers/rf-transceiver/hmc630x/hmc630x.c
Comment (buha) [reply drivers/rf-transceiver/hmc630x/hmc630x.c:532]: yes
https://github.com/analogdevicesinc/no-OS/commit/7d3a8eaf416ee6f8ef254a3b68a6b436f4313307#diff-63ee9393e76ad03b20befef3fd69217950306773e724fea402048e5e7bcd27fdR137

and dev->type can't be changed by API user

**Comment by buha** [comment]
V8:
- fixed some formatting issues

**Review by amiclaus — APPROVED** [review]

**Pull request merged**
